x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/BinaryDateTimeProcessor.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.sql.expression.function.scalar.datetime;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xpack.sql.common.io.SqlStreamInput;
+import org.elasticsearch.xpack.sql.expression.gen.processor.BinaryProcessor;
+import org.elasticsearch.xpack.sql.expression.gen.processor.Processor;
+
+import java.io.IOException;
+import java.time.ZoneId;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.sql.expression.function.scalar.datetime.BinaryDateTimeProcessor.BinaryDateOperation.TRUNC;
+
+public abstract class BinaryDateTimeProcessor extends BinaryProcessor {
+
+    public enum BinaryDateOperation {
+        TRUNC,
+        PART;
+    }
+
+    private final ZoneId zoneId;
+
+    public BinaryDateTimeProcessor(Processor source1, Processor source2, ZoneId zoneId) {
+        super(source1, source2);
+        this.zoneId = zoneId;
+    }
+
+    public BinaryDateTimeProcessor(StreamInput in) throws IOException {
+        super(in);
+        zoneId = SqlStreamInput.asSqlStream(in).zoneId();
+    }
+
+    @Override
+    protected void doWrite(StreamOutput out) {
+    }
+
+    ZoneId zoneId() {
+        return zoneId;
+    }
+
+    @Override
+    protected abstract Object doProcess(Object left, Object right);
+
+    public static BinaryDateTimeProcessor asProcessor(BinaryDateOperation operation, Processor left, Processor right, ZoneId zoneId) {
+        if (operation == TRUNC) {
```

Comment:
Same comment here. Why not delegating the choice of which processor to return to the subclasses themselves, as they are the decision makers. What if this abstract BinaryDateTimeProcessor will have another subclass in the future? It's weird to come back to the base class and change its implementation based on new subclasses that were added. Rather the subclasses should provide its specific implementation of a certain aspect of code and the base class should remain unchanged.

Reply:
Since this a static method (used to avoid subclassing the BinaryDateTimePipe) I chose the enum. I totally agree with your suggestion, but since a 3rd argument will be added to the DATE_PART function and 2 more 3 args function will be added (DATE_ADD/DIFF), I'd postpone this refactoring until the 3rd arg is added (separate PR) and the layout of the classes will be settled. I'm adding a TODO to as a reminder to remove the enum.